### PR TITLE
Scio REPL

### PR DESCRIPTION
**Pull request by ravwojdyla**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -275,6 +275,23 @@ lazy val scioExamples: Project = Project(
   scioTest % "test"
 )
 
+lazy val scioRepl: Project = Project(
+  "scio-repl",
+  file("scio-repl"),
+  settings = commonSettings ++ Seq(
+    libraryDependencies <++= (scalaVersion) { scalaVersion => Seq(
+      "jline" % "jline" % scalaVersion.take(4),
```
Comment (nevillelyh) [inline build.sbt]: use `scalaBinaryVerion`?

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 2]
```diff
@@ -132,12 +143,46 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: O
...
+
+  /** *
+    * Add artifact to stage in Dataflow - artifact can be jar/text-files etc.
+    * NOTE: currently one can add artifacts only before pipeline object is created
+    * @param extraLocalArtifacts extra local artifacts
+    * @return [[ScioContext]]
+    */
+  def addArtifacts(extraLocalArtifacts: List[String]): ScioContext = {
+    require(_pipeline == null, "Pipeline object already created - can't add artifacts anymore!")
+    artifacts ++ extraLocalArtifacts
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: This statement has no effect?

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 3]
```diff
@@ -132,12 +143,46 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: O
   private val _promises: Buffer[(Promise[AnyRef], AnyRef)] = Buffer.empty
   private val _bigQueryJobs: Buffer[JobReference] = Buffer.empty
   private val _accumulators: MSet[String] = MSet.empty
+  private val _filesToStage = List[String]()
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: I don't see where this is used?

**Comment by ravwojdyla** [comment]
Initial REPL implementation.

To start:

    sbt 'project scio-repl' assembly

For local tests:

    java -jar scio-repl/target/scala-2.11/scio-repl-0.1.1-SNAPSHOT-fat.jar
    scio> :newScio
    scio> sc.textFile("README.md").map( _ + "mod").saveAsTextFile("/tmp/modReadme")
    scio> :runScio

For distributed test:

    java -jar scio-repl/target/scala-2.11/scio-repl-0.1.1-SNAPSHOT-fat.jar --project=<project-id> --stagingLocation=<staging-dir> --zone=<zone> --runner=<runner>
    scio> :newScio rsc
    scio> rsc.textFile("gs://<gs files>").map( _ + "mod4" ).saveAsTextFile("gs://<output-dir>")
    scio> :runScio rsc
    

Still to do:

- [ ] add readme and examples
- [ ] add async executions
- [ ] add implicits to ease handling SCollections, Taps and Futures

**Comment by nevillelyh** [comment]
Does scio-repl cross compile to 2.10? Use SBT `+compile` command.

**Comment by ravwojdyla** [comment]
@nevillelyh there are some issues with `2.10`.

**Comment by ravwojdyla** [comment]
@nevillelyh should compile with `2.10` now - but there is an issue inside the REPL with `JLine`:
    
    Failed to created JLineReader: java.lang.NoClassDefFoundError: scala/tools/jline/console/completer/Completer

**Comment by ravwojdyla** [comment]
@nevillelyh and should be fine now on both `2.10` and `2.11` + fixes for previous comments

File: scio-repl/src/main/scala-2.11/com/spotify/scio/repl/ILoopCompat.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import java.io.BufferedReader
+
+import scala.tools.nsc.interpreter.{ILoop, JPrintWriter}
+
+/**
+ * ILoop for ScioLoop
+ */
+class ILoopCompat(in: Option[BufferedReader], out: JPrintWriter)
+  extends ILoop(in, out) {
+  def addThunk(f: => Unit): Unit = intp.initialize(f)
```
Comment (nevillelyh) [inline scio-repl/src/main/scala-2.11/com/spotify/scio/repl/ILoopCompat.scala]: This looks the same as the 2.10 version except for the `override`?

File: build.sbt [hunk 5]
```diff
@@ -275,6 +278,29 @@ lazy val scioExamples: Project = Project(
...
+  "scio-repl",
+  file("scio-repl"),
+  settings = commonSettings ++ Seq(
+    libraryDependencies ++= Seq(
+      "jline" % "jline" % scalaBinaryVersion.value,
+      "org.scala-lang" % "scala-compiler" % scalaVersion.value,
+      "org.scala-lang" % "scala-reflect" % scalaVersion.value
+    ),
+    libraryDependencies ++= (
+      if (scalaVersion.value.startsWith("2.10"))
```
Comment (nevillelyh) [inline build.sbt]: Use `scalaBinaryVersion` as well?

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 6]
```diff
@@ -43,18 +43,20 @@ import com.spotify.scio.values._
...
 import scala.reflect.ClassTag
+import scala.util.Try
 
 /** Convenience object for creating [[ScioContext]] and [[Args]]. */
 object ContextAndArgs {
   /** Create [[ScioContext]] and [[Args]] for command line arguments. */
   def apply(args: Array[String]): (ScioContext, Args) = {
     val (_opts, _args) = ScioContext.parseArguments[DataflowPipelineOptions](args)
-    (new ScioContext(_opts, _args.optional("testId")), _args)
+    (new ScioContext(_opts, List[String](), _args.optional("testId")), _args)
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: Minor: I'll use Nil for empty lists.

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 7]
```diff
@@ -103,16 +111,19 @@ object ScioContext {
  * @groupname Ungrouped Other Members
  */
 // scalastyle:off number.of.methods
-class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: Option[String]) {
+class ScioContext private[scio] (val options: DataflowPipelineOptions, private var artifacts: List[String], testId: Option[String]) {
 
   import Implicits._
 
+  private val LOG = LoggerFactory.getLogger(ScioContext.getClass)
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: Minor: it's called `logger` everywhere else in the scala code base. Better be consistent but what should we use?

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+          case v if v.nonEmpty => intp.interpret ("val " + scioContextVal + " = ScioContext(__scio__df__opts__, List(\"" + nextReplJar + "\"))")
+          case _ => intp.interpret ("val sc = ScioContext(__scio__df__opts__, List(\"" + nextReplJar + "\"))")
+        }
+      })
+      Result.default
+    })
+  private val getResultCmd = LoopCommand.cmd("runScio", ":runScio <[scio-context-val-name] | sc>",  "run Scio pipeline",
+    (scioContextVal) => {
+      scioClassLoader.get.createReplCodeJar
+      val scioResult = scioContextVal match {
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala]: scioContextVal is a string right? Match/case is overkill and if/else is easier to read?

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import scala.tools.nsc.interpreter.{IR, JPrintWriter}
+
+/**
+ * ScioILoop - core of Scio REPL
+ * @param scioClassLoader [[ScioReplClassLoader]] used for runtime classloading
+ * @param args user arguments for Scio REPL
+ * @param in [[ILoopCompat]] input
+ * @param out [[ILoopCompat]] output
+ */
+class ScioILoop(scioClassLoader: Option[ScioReplClassLoader],
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala]: Why `Option`? Don't see a None case.

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  /**
+   * Gets the list of commands that this REPL supports.
+   *
+   * @return a list of the command supported by this REPL.
+   */
+  override def commands: List[LoopCommand] = super.commands ++ scioCommands
+
+  protected def imports: List[String] = List(
+    //TODO: add implicites etc
+    "com.spotify.scio.ScioContext",
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala]: `com.spotify.scio._` instead?

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+  /**
+   * Creates a jar file in a temporary directory containing the code thus far compiled by the REPL.
+   *
+   * @return some file for the jar created, or `None` if the REPL is not running.
+   */
+  //private[scio]
+  def createReplCodeJar: String = {
+    scioREPL.map { repl =>
+      //scala 2.11 : use repl.replOutput.dir
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala]: // TODO so it's more greppable in the future?

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+class ScioReplClassLoader(urls: Array[URL], parent: ClassLoader, detachedParent: ClassLoader)
+  extends URLClassLoader(urls, parent) {
+  private val LOG = LoggerFactory.getLogger(classOf[ScioReplClassLoader])
+
+  private val replJarName = "scio-repl-session.jar"
+  private var nextReplJarDir: File = genNextReplCodeJarDir
+
+  private var scioREPL: Option[ILoop] = None
+
+  def setRepl(repl: Option[ILoop]) = scioREPL = repl
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala]: `:Unit = scioREPL = repl` so it's less confusing

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        LOG.debug("Found " + name + " in REPL classloader " + scioREPL.get.classLoader.toString)
+        return clazz.get
+      }
+    }
+    super.loadClass(name)
+  }
+
+  def genNextReplCodeJarDir: File = Files.createTempDir()
+
+  def getNextReplCodeJarPath: String = {
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala]: I usually omit {} for one line methods, even if it's on a separate line.

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 14]
```diff
@@ -138,6 +149,39 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: O
     new SCollectionImpl[T](p, this)
 
   // =======================================================================
+  // Extra artifacts - jars/files etc
+  // =======================================================================
+
+  /** Borrowed from Dataflow */
+  private def detectClassPathResourcesToStage(classLoader: ClassLoader) = {
+    require(classLoader.isInstanceOf[URLClassLoader], "Current ClassLoader is '" + classLoader + "' only URLClassLoaders are supported")
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: Line break/wrap for readability

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 15]
```diff
@@ -138,6 +149,39 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: O
...
+  /** *
+    * Add artifact to stage in Dataflow - artifact can be jar/text-files etc.
+    * NOTE: currently one can add artifacts only before pipeline object is created
+    * @param extraLocalArtifacts extra local artifacts
+    * @return [[ScioContext]]
+    */
+  def addArtifacts(extraLocalArtifacts: List[String]): ScioContext = {
+    require(_pipeline == null, "Pipeline object already created - can't add artifacts anymore!")
+    artifacts ++= extraLocalArtifacts
+    this
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: `: Unit` for consistency.

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 16]
```diff
@@ -138,6 +149,39 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: O
...
+  }
+
+  /** Compute list of files to stage in dataflow */
+  private def getFilesToStage(extraLocalArtifacts: List[String]): List[String] = {
+    val finalLocalArtifacts = detectClassPathResourcesToStage(classOf[DataflowPipelineRunner].getClassLoader) ++ extraLocalArtifacts
+    //TODO: add logging, stats
+    finalLocalArtifacts.toList
+  }
+
+  /** *
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: In IntelliJ pref, Editor -> Code Style -> Scala -> ScalaDoc, untick Use ScalaDoc indent for leading asterisk.

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 17]
```diff
@@ -138,6 +149,39 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions, testId: O
...
+    * Add artifact to stage in Dataflow - artifact can be jar/text-files etc.
+    * NOTE: currently one can add artifacts only before pipeline object is created
+    * @param extraLocalArtifacts extra local artifacts
+    * @return [[ScioContext]]
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: No need to document obvious params, retvals, etc.

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+  private val getNextJarCmd = LoopCommand.nullary("nextJar", "gets next path of Scio REPL jar",
+    () => { Result.resultFromString(scioClassLoader.get.getNextReplCodeJarPath) } )
+  private val getNewScioContextCmd = LoopCommand.cmd("newScio", "newScio <[scio-context-val-name] | sc>", "gets new Scio context",
+    (scioContextVal) => {
+      val nextReplJar = scioClassLoader.get.getNextReplCodeJarPath
+      intp.beQuietDuring({
+        intp.interpret("val __scio__opts__ = " + args.map('"' + _ + '"'))
+        intp.interpret("val __scio__df__opts__ = PipelineOptionsFactory.fromArgs(__scio__opts__.toArray).as(classOf[DataflowPipelineOptions])")
+
+        scioContextVal match {
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala]:
```
val scName = if (scioContextVal.nonEmpty) scioContextVal else "sc"
...
```

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    (scioContextVal) => {
+      scioClassLoader.get.createReplCodeJar
+      val scioResult = scioContextVal match {
+        case v if !v.isEmpty => intp.interpret(v + ".close")
+        case _ => intp.interpret("sc.close")
+      }
+      Result.default
+    })
+
+  private val scioCommands: List[LoopCommand] = List( createJarCmd, getNextJarCmd, getNewScioContextCmd, getResultCmd)
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala]: Extra space in `( createJarCmd`

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+  override def printWelcome() {
+    echo("Welcome to Scio REPL!")
+  }
+
+  private val createJarCmd = LoopCommand.nullary("createJar", "creates Scio REPL runtime jar",
+    () => { Result.resultFromString(scioClassLoader.get.createReplCodeJar) } )
+  private val getNextJarCmd = LoopCommand.nullary("nextJar", "gets next path of Scio REPL jar",
+    () => { Result.resultFromString(scioClassLoader.get.getNextReplCodeJarPath) } )
+  private val getNewScioContextCmd = LoopCommand.cmd("newScio", "newScio <[scio-context-val-name] | sc>", "gets new Scio context",
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioILoop.scala]: Line wrap

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    new File(nextReplJarDir, replJarName).getAbsolutePath
+  }
+
+  /**
+   * Creates a jar file in a temporary directory containing the code thus far compiled by the REPL.
+   *
+   * @return some file for the jar created, or `None` if the REPL is not running.
+   */
+  //private[scio]
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala]: debug comments

File: scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+        scioREPL.map( _.classLoader.loadClass(name))
+      } catch {
+        case e: Exception => {
+          LOG.error("Could not find " + name + " in REPL classloader", e)
+          null
+        }
+      }
+      if (clazz.isDefined) {
+        LOG.debug("Found " + name + " in REPL classloader " + scioREPL.get.classLoader.toString)
+        return clazz.get
```
Comment (nevillelyh) [inline scio-repl/src/main/scala/com/spotify/scio/repl/ScioReplClassLoader.scala]: No `return` here and `else { super.loadClass } later`

File: scio-repl/src/main/scala-2.11/com/spotify/scio/repl/ILoopCompat.scala (hunk 4 above)
Comment (ravwojdyla) [inline scio-repl/src/main/scala-2.11/com/spotify/scio/repl/ILoopCompat.scala]: Yep - has to - discussed IRL.

File: build.sbt (hunk 5 above)
Comment (ravwojdyla) [inline build.sbt]: Fixed + bigquery project if too.

**Comment by ravwojdyla** [comment]
@nevillelyh this is read for final review and merge if fine, Would be great to check if `readme` works fine for you. We can work on more magic/implicits once we have more feedback from users.

**Comment by nevillelyh** [comment]
Duplicate with #22

**Pull request closed without merging**